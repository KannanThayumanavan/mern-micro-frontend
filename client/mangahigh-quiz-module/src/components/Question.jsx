import React from 'mangahigh-libs/react';
import Button from 'mangahigh-components/Button';
import MultiOptions from 'mangahigh-components/MultiOptions';
import PropTypes from 'prop-types';

const Question = ({ 
	question: {
		id,
		question,
		options,
	},
	getSelectedValue,
	showSubmission, 
	updateProgress,
}) => (
	<div className="p-3">
		<h5>{question}</h5>
		<MultiOptions 
	    options={options} 
	    groupName={id}
	    getSelectedValue={getSelectedValue} 
  	/>
  	<div className="p-3">
  	{showSubmission
  		? (
  				<Button 
						type='primary' 
						buttonLabel='Submit' 
						buttonOnClick={() => updateProgress('result')}
					/>
  			)
  		: (
  				<Button 
						type='primary' 
						buttonLabel='Next'
						buttonOnClick={() => updateProgress('quiz')}
					/>					
  			)
  	}
  	</div>
		
	</div>
);

Question.propTypes = {
	question: PropTypes.shape({
		id: PropTypes.string.isRequired,
		question: PropTypes.string.isRequired,
		options: PropTypes.arrayOf(PropTypes.string).isRequired,
	}),
	getSelectedValue: PropTypes.func.isRequired,
	showSubmission: PropTypes.bool.isRequired,
	updateProgress: PropTypes.func.isRequired,
};

export default Question;