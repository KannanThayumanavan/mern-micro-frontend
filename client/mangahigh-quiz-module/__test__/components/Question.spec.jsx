import React from 'react';
import { shallow } from 'enzyme';
import Question from '../../src/components/Question';

jest.mock('mangahigh-libs/react', () => jest.requireActual('react'),
	{ virtual: true }
);

jest.mock('mangahigh-components/Button', () => 'Button'			
, { virtual: true });

jest.mock('mangahigh-components/MultiOptions', () => 'MultiOptions'			
, { virtual: true });

describe('Test Question Component', () => {
	const mockCallback = jest.fn();
	const props = {
		question: {
			id: 'test_id',
			question: 'test question 1',
			options: ['test option 1', 'test option 2'],
		},
		getSelectedValue: jest.fn(),
		showSubmission: false, 
		updateProgress: mockCallback,
	};
	
	describe('Test navigation between quiz questions', () => {
		it('should render correctly with required props', () => {
			const renderedModule = shallow(<Question {...props} />);
	    expect(renderedModule).toMatchSnapshot();
		});

		it('should callback to change question on click of Next button', () => {
			const renderedModule = shallow(<Question {...props} />);
			renderedModule.find('Button').props().buttonOnClick();
			expect(mockCallback).toHaveBeenCalledWith('quiz');
		});
	});
	
	describe('Test quiz submission and result section', () => {
		const endQuizProp = {
			...props,
			showSubmission: true, 
		};
		it('should show Submit button at end of Quiz', () => {			
			const renderedModule = shallow(<Question {...endQuizProp} />);		
	    expect(renderedModule).toMatchSnapshot();
		});

		it('should callback to play again on click of Play again button', () => {
			const renderedModule = shallow(<Question {...endQuizProp} />);
			renderedModule.find('Button').props().buttonOnClick();
			expect(mockCallback).toHaveBeenCalledWith('result');
		});
	});
});