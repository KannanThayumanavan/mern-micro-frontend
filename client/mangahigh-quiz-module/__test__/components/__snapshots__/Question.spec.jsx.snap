// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Question Component Test navigation between quiz questions should render correctly with required props 1`] = `
<div
  className="p-3"
>
  <h5>
    test question 1
  </h5>
  <MultiOptions
    getSelectedValue={[MockFunction]}
    groupName="test_id"
    options={
      Array [
        "test option 1",
        "test option 2",
      ]
    }
  />
  <div
    className="p-3"
  >
    <Button
      buttonLabel="Next"
      buttonOnClick={[Function]}
      type="primary"
    />
  </div>
</div>
`;

exports[`Test Question Component Test quiz submission and result section should show Submit button at end of Quiz 1`] = `
<div
  className="p-3"
>
  <h5>
    test question 1
  </h5>
  <MultiOptions
    getSelectedValue={[MockFunction]}
    groupName="test_id"
    options={
      Array [
        "test option 1",
        "test option 2",
      ]
    }
  />
  <div
    className="p-3"
  >
    <Button
      buttonLabel="Submit"
      buttonOnClick={[Function]}
      type="primary"
    />
  </div>
</div>
`;
